readthedocs/proxito/views/utils.py

```diff
@@ -19,6 +19,24 @@ def fast_404(request, *args, **kwargs):
     return HttpResponse('Not Found.', status=404)
 
 
+def proxito_404_page_handler(request, exception=None, template_name='404.html'):
+    """
+    Decide what 404 page return depending if it's an internal NGINX redirect.
+
+    We want to return fast when the 404 is used as an internal NGINX redirect to
+    reach our ``ServeError404`` view. However, if the 404 exception was risen
+    inside ``ServeError404`` view, we want to render the default Read the Docs
+    Maze page.
+    """
+
+    if request.resolver_match.url_name != 'proxito_404_handler':
+        return fast_404(request, exception, template_name)
```

Comment:
When would this ever be True? It's only configured on our proxito views, so it seems like we could just use the normal RTD 404 handler here, no?

Reply:
If you hit a URL of a non existing project (*) there are going to be two 404 happening there:

1. internal NGINX redirect, produced at decorator project mapping: https://github.com/readthedocs/readthedocs.org/blob/804887591a0882b0f6d3603a0034b5cc12e79755/readthedocs/proxito/views/decorators.py#L76
2. user-facing 404, in our ServeError404 view itself: https://github.com/readthedocs/readthedocs.org/blob/804887591a0882b0f6d3603a0034b5cc12e79755/readthedocs/proxito/views/serve.py#L280

The 1) is going to execute `fast_404` and 2) is going to execute the user-facing one.

We could use always the user-facing one, but we "loose the speed" of the `fast_404` in those cases. I'm fine with that, since it's simpler, though.

(*) any 404 happening inside the `ServeDocs` should fastly respond 404 so it's cached by our own `ServeError404` handler. In the end, the last 404 which will be shown to the user must be the one that calls `render(template)`, all the others could be the fast one.